src/DotNetty.Handlers/Timeout/ReadTimeoutHandler.cs

```diff
@@ -0,0 +1,227 @@
+                // and initialization will occur there.
+            }
+        }
+
+        public override void HandlerRemoved(IChannelHandlerContext context)
+        {
+            this.Destroy();
+        }
+
+        public override void ChannelRegistered(IChannelHandlerContext context)
+        {
+            // Initialize early if channel is active already.
+            if (context.Channel.Active)
+            {
+                this.Initialize(context);
+            }
+
+            base.ChannelRegistered(context);
+        }
+
+        public override void ChannelActive(IChannelHandlerContext context)
+        {
+            // This method will be invoked only if this handler was added
+            // before channelActive() event is fired.  If a user adds this handler
+            // after the channelActive() event, initialize() will be called by beforeAdd().
+            this.Initialize(context);
+            base.ChannelActive(context);
+        }
+
+        public override void ChannelInactive(IChannelHandlerContext context)
+        {
+            this.Destroy();
+            base.ChannelInactive(context);
+        }
+
+        public override void ChannelRead(IChannelHandlerContext context, object message)
+        {
+            this.reading = true;
+            context.FireChannelRead(message);
+        }
+
+        public override void ChannelReadComplete(IChannelHandlerContext context)
+        {
+            Interlocked.Exchange(ref this.lastReadTime, TimeUtil.GetSystemNanoTime());
+            this.reading = false;
+            context.FireChannelReadComplete();
+        }
+
+        private void Initialize(IChannelHandlerContext context)
+        {
+            switch(this.state)
+            {
+                case 1:
+                case 2:
+                    return;
+            }
+
+            this.state = 1;
+
+            this.lastReadTime = TimeUtil.GetSystemNanoTime();
+            if (this.timeoutNanos > 0) 
+            {
+                this.timeout = context.Executor.Schedule(this.MakeReadTimeOutTask(context), 
+                    TimeUtil.GetTimeSpanFromNanos(this.timeoutNanos));
+            }
+        }
+
+        private void Destroy()
+        {
+            state = 2;
+
+            if(this.timeout != null)
+            {
+                this.timeout.Cancel();
+                this.timeout = null;
+            }
+        }
+
+        /// <summary>
+        /// Is called when a read timeout was detected.
+        /// </summary>
+        /// <param name="context">Context.</param>
+        protected void ReadTimedOut(IChannelHandlerContext context)
+        {
+            if(!this.closed)
+            {
+                context.FireExceptionCaught(ReadTimeoutException.Instance);
+                context.Flush();
+                this.closed = true;
+            }
+        }
+
+        Action MakeReadTimeOutTask(IChannelHandlerContext context)
```

Comment:
you can add a `static readonly Action<object, object> ReadTimeoutAction = HandleReadTimeout;`
and the method as follows:
```
static void HandleReadTimeout(object handler, object ctx)
{
    var self = (ReadTimeoutHandler)handler; // instead of this
    var context = (IChannelHandlerContext)ctx;
    ... 
}
```

Then to call schedule like so: `Executor.Schedule(ReadTimeoutAction, this, context, this.timeout)`. Saves 2 allocations per call (delegate and closure object).